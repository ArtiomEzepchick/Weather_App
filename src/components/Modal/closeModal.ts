export const closeModal = (delay: number = 500): void => {
  const overlay = document.querySelector(".overlay");

  if (overlay) {
    if (overlay.classList.contains("show")) {
      overlay.classList.add("hidden");

      setTimeout(() => {
        overlay.classList.remove("show");
      }, delay);
    }
  }
};
