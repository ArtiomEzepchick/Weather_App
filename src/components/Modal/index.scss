@import "../../styles/variables.scss";

.modal-container {
  display: block;
  width: 100%;
  height: 100%;

  .modal-window {
    display: flex;
    justify-content: space-around;
    align-items: center;
    flex-direction: column;
    position: fixed;
    width: 25rem;
    height: $size-s;
    border-radius: 2rem;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    padding: 2rem;
    z-index: 100;
    opacity: 1;
    box-shadow: 24px 19px 8px 0px rgba(34, 60, 80, 0.2);
    background-color: #ff5f5fbb;
    font-size: 1.2rem;
    text-align: center;

    .modal-content {
      display: flex;
      flex-direction: column;
      gap: 1rem;
    }

    .modal-buttons {
      display: flex;
      gap: 1rem;
      button {
        min-width: 5rem;

        &:hover {
          border-color: $color-red;
          color: $color-red;
        }
      }
    }
  }

  @media (max-width: $window-s) {
    .modal-window {
      width: $size-m;
      height: $size-l;
    }
  }
}
