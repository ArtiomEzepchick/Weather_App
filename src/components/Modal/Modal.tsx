import React, { useRef, useEffect, useCallback, RefObject } from "react"
import { Button, InputRef } from "antd"
import { createPortal } from "react-dom"
import { useDispatch } from "react-redux"
import { Dispatch } from "redux"

import Overlay from "../Overlay/Overlay"

import { useScrollLock } from "../../hooks/useScrollLock"
import { clearError, setInputCityValue } from "../../model/weather/actions/actions"
import { closeModal } from "./closeModal"

import "./index.scss"

type Props = {
    contentText: string;
    isModalOpen: boolean;
    inputRef: RefObject<InputRef>;
}

const Modal: React.FC<Props> = ({
    contentText,
    isModalOpen,
    inputRef
}) => {
    const modalRef = useRef<HTMLDivElement>(null)
    const { lockScroll, unlockScroll } = useScrollLock()
    const modalRoot = document.getElementById('modal') as HTMLElement
    const dispatch: Dispatch = useDispatch()

    const handleCloseModal = useCallback((): void => {
        closeModal()
        unlockScroll()

        setTimeout(() => {
            dispatch(clearError())
        }, 200)
    }, [dispatch, unlockScroll])

    const handleCloseModalAndEdit = useCallback((): void => {
        if (inputRef.current) inputRef.current.focus()

        closeModal()
        unlockScroll()

        setTimeout(() => {
            dispatch(clearError())
            dispatch(setInputCityValue(''))
        }, 200)
    }, [dispatch, inputRef, unlockScroll])

    useEffect(() => {
        isModalOpen ? lockScroll() : unlockScroll()

        const handleClickOutside = (e: MouseEvent): void => {
            const target = e.target as HTMLDivElement
            if (modalRef.current && !modalRef.current.contains(target)) {
                handleCloseModal()
            }
        }

        document.addEventListener("mousedown", handleClickOutside)

        return () => document.removeEventListener("mousedown", handleClickOutside)
    }, [
        modalRef,
        isModalOpen,
        lockScroll,
        unlockScroll,
        handleCloseModal
    ])

    return (
        createPortal(
            <Overlay isModalOpen={isModalOpen}>
                <div className="modal-container">
                    <div
                        className="modal-window"
                        ref={modalRef}
                    >
                        <section className="modal-content">
                            <h2>An error has occurred</h2>
                            <span>{contentText}</span>
                        </section>
                        <section className="modal-buttons">
                            <Button onClick={handleCloseModal}>Ok</Button>
                            <Button onClick={handleCloseModalAndEdit}>Return to edit</Button>
                        </section>
                    </div>
                </div>
            </Overlay>, modalRoot
        )
    )
}

export default Modal