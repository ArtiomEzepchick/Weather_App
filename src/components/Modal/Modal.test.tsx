import React, { ReactNode, ReactPortal } from "react"
import ReactDOM from "react-dom"
import { render, screen, fireEvent } from "@testing-library/react"
import '@testing-library/jest-dom'

import Modal from "./Modal"

describe('Modal', () => {
    const oldCreatePortal = ReactDOM.createPortal
    const inputRef = { current: null }
    const contentText = 'Content'

    const propsModalOpen = {
        isModalOpen: true,
        dispatch: jest.fn(),
        contentText,
        inputRef
    }

    const propsModalClosed = {
        ...propsModalOpen,
        isModalOpen: false
    }

    beforeAll(() => {
        ReactDOM.createPortal = (node: ReactNode): ReactPortal => node as ReactPortal
    })

    afterAll(() => {
        ReactDOM.createPortal = oldCreatePortal
    })

    it('renders correctly', () => {
        render(<Modal { ...propsModalOpen } />)

        const overlay = screen.getByTestId('overlay')
        const modalHeader = screen.getByTestId('modalHeader')
        const modalContent = screen.getByTestId('modalContent')

        expect(overlay).toHaveClass('show')
        expect(modalHeader).toHaveTextContent('An error has occurred')
        expect(modalContent).toHaveTextContent(contentText)
    })

    it('handles onClick close modal', () => {
        const { rerender } = render(<Modal { ...propsModalOpen } />)

        const overlay = screen.getByTestId('overlay')
        const button = screen.getByText('Ok')

        expect(overlay).toHaveClass('show')
        fireEvent.click(button)
        expect(overlay).toHaveClass('hidden')

        rerender(<Modal { ...propsModalClosed }/>)

        expect(overlay).not.toHaveClass('hidden')
        expect(overlay).not.toHaveClass('show')
    })

    it('handles onClick close modal and edit', () => {
        const { rerender } = render(<Modal { ...propsModalOpen } />)

        const overlay = screen.getByTestId('overlay')
        const button = screen.getByText('Return to edit')
        
        expect(overlay).toHaveClass('show')
        fireEvent.click(button)
        expect(overlay).toHaveClass('hidden')

        rerender(<Modal { ...propsModalClosed } />)

        expect(overlay).not.toHaveClass('hidden')
        expect(overlay).not.toHaveClass('show')
    })
})