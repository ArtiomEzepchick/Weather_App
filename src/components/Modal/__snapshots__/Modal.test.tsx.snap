// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal handles onClick close modal 1`] = `
<div
  class="overlay"
  data-testid="overlay"
>
  <div
    class="modal-container"
    data-testid="modalContainer"
  >
    <div
      class="modal-window"
    >
      <section
        class="modal-content"
      >
        <h2
          data-testid="modalHeader"
        >
          An error has occurred
        </h2>
        <p
          data-testid="modalContent"
        >
          Content
        </p>
      </section>
      <section
        class="modal-buttons"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-1vtf12y ant-btn-default"
          type="button"
        >
          <span>
            Ok
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-1vtf12y ant-btn-default"
          type="button"
        >
          <span>
            Return to edit
          </span>
        </button>
      </section>
    </div>
  </div>
</div>
`;

exports[`Modal handles onClick close modal and edit 1`] = `
<div
  class="overlay"
  data-testid="overlay"
>
  <div
    class="modal-container"
    data-testid="modalContainer"
  >
    <div
      class="modal-window"
    >
      <section
        class="modal-content"
      >
        <h2
          data-testid="modalHeader"
        >
          An error has occurred
        </h2>
        <p
          data-testid="modalContent"
        >
          Content
        </p>
      </section>
      <section
        class="modal-buttons"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-1vtf12y ant-btn-default"
          type="button"
        >
          <span>
            Ok
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-1vtf12y ant-btn-default"
          type="button"
        >
          <span>
            Return to edit
          </span>
        </button>
      </section>
    </div>
  </div>
</div>
`;

exports[`Modal renders correctly 1`] = `
<div
  class="overlay show"
  data-testid="overlay"
>
  <div
    class="modal-container"
    data-testid="modalContainer"
  >
    <div
      class="modal-window"
    >
      <section
        class="modal-content"
      >
        <h2
          data-testid="modalHeader"
        >
          An error has occurred
        </h2>
        <p
          data-testid="modalContent"
        >
          Content
        </p>
      </section>
      <section
        class="modal-buttons"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-1vtf12y ant-btn-default"
          type="button"
        >
          <span>
            Ok
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-1vtf12y ant-btn-default"
          type="button"
        >
          <span>
            Return to edit
          </span>
        </button>
      </section>
    </div>
  </div>
</div>
`;
